### ISSUE TITLE
OrderedDict fails to infer types of generators.

### ISSUE THREAD

**Issue by bicycle1885** [JuliaCollections/DataStructures.jl#248]
On 3afb7a24c304c67a0b3d432af89a2550333eb071, it fails to infer element types as follows:
```
julia> typeof(Dict(x => 2x for x in [1,2,3]))  # this works
Dict{Int64,Int64}

julia> typeof(OrderedDict(x => 2x for x in [1,2,3]))  # but this doesn't
DataStructures.OrderedDict{Any,Any}
```

I'm using Julia 0.5.0.

**Comment by oxinabox** [comment]
This is because julia does not do inference on the types of generators.
This is basically an issue for JuliaLang/julia.
I believe there is a good reason not to.
Like that it can't always be done.
And so the behavior of the code should not depend on it.

The `eltype` for generators is not set:
```
julia> Base.iteratoreltype(typeof(x=>2x for x in [1,2,3]))
Base.EltypeUnknown()

julia> eltype(x=>2x for x in [1,2,3])
Any
```

One could force it, via something like

```
force_eltype(::Type{T}) where {T} = eltype(T) #fallback
force_eltype(gen::T) where {T} =force_eltype(T)

function force_eltype(::Type{Base.Generator{B,F}}) where{B,F}
    back_et = force_eltype(B)
    gen_ets = Base.return_types(F.instance, (back_et,))
    Union{gen_ets...}
end
```

But I'm sure there is a good reason not to.

**Comment by oxinabox** [comment]
It seems we are already doing something like this with `Base._default_eltype` for accumulators.
So maybe this is Ok

**Comment by cstjean** [comment]
I don't like using `Base._default_eltype` either, but if Base cheats for its `Dict` implementation, IMHO DataStructures can/should cheat too, if only to provide a consistent user experience. There are some [comments on Discourse about it](https://discourse.julialang.org/t/can-eltype-deduce-the-element-type-of-a-generator/6429/13), apparently the `eltype` of generators is a contentious topic.

**Comment by oxinabox** [comment]
I agree.

**Comment by laborg** [comment]
This is fixed by now (Julia 1.9.3 and DataStructures 0.18.5:
```
julia> typeof(Dict(x => 2x for x in [1,2,3]))  
Dict{Int64, Int64}

julia> typeof(OrderedDict(x => 2x for x in [1,2,3])) 
OrderedDict{Int64, Int64}
```
